networkit/cpp/components/ConnectedComponentsImpl.cpp

```diff
@@ -0,0 +1,152 @@
+
+    std::queue<node> q;
+
+    // perform breadth-first searches
+    for (const node u : G->nodeRange()) {
+        if (component[u] != none)
+            continue;
+
+        component.setUpperBound(numComponents + 1);
+        const index c = numComponents;
+
+        q.push(u);
+        component[u] = c;
+
+        do {
+            const node v = q.front();
+            q.pop();
+            ++visitedNodes;
+
+            const auto visitNeighbor = [&](node w) -> void {
+                if (component[w] == none) {
+                    q.push(w);
+                    component[w] = c;
+                }
+            };
+
+            // enqueue neighbors, set component
+            G->forNeighborsOf(v, visitNeighbor);
+            if (weaklyCC)
+                G->forInNeighborsOf(v, visitNeighbor);
+
+        } while (!q.empty());
+
+        ++numComponents;
+
+        if (visitedNodes == G->numberOfNodes())
+            break;
+    }
+
+    hasRun = true;
+}
+
+Partition ConnectedComponentsImpl::getPartition() const {
+    assureFinished();
+    return this->component;
+}
+
+std::vector<std::vector<node>> ConnectedComponentsImpl::getComponents() const {
+    assureFinished();
+
+    std::vector<std::vector<node>> result(numComponents);
+    G->forNodes([&](node u) { result[component[u]].push_back(u); });
+
+    return result;
+}
+
+std::map<index, count> ConnectedComponentsImpl::getComponentSizes() const {
+    assureFinished();
+    return this->component.subsetSizeMap();
+}
+
+Graph ConnectedComponentsImpl::extractLargestConnectedComponent(bool compactGraph) const {
+    if (G->isEmpty())
+        return *G;
+
+    ConnectedComponentsImpl cc(*G, weaklyCC);
+    cc.run();
+
+    const auto compSizes = cc.getComponentSizes();
+    if (compSizes.size() == 1) {
+        if (compactGraph)
+            return GraphTools::getCompactedGraph(*G, GraphTools::getContinuousNodeIds(*G));
+        return *G;
+    }
+
+    const auto largestCC = std::max_element(
+        compSizes.begin(), compSizes.end(),
+        [](const std::pair<index, count> &x, const std::pair<index, count> &y) -> bool {
+            return x.second < y.second;
+        });
+
+    if (compactGraph) {
+        std::unordered_map<node, node> continuousNodeIds;
+        index nextId = 0;
+        G->forNodes([&](const node u) {
+            if (cc.componentOfNode(u) == largestCC->first)
+                continuousNodeIds[u] = nextId++;
+        });
+
+        return GraphTools::getRemappedGraph(
+            *G, largestCC->second, [&](const node u) { return continuousNodeIds[u]; },
+            [&](const node u) { return cc.componentOfNode(u) != largestCC->first; });
+
+    } else {
+        Graph S(*G);
+        const auto components = cc.getComponents();
+        for (size_t i = 0; i < components.size(); ++i) {
+            if (i != largestCC->first)
+                for (const node u : components[i])
+                    S.removeNode(u);
+        }
```

Comment:
If we already have the connected components as sets of nodes, I wonder why `GraphTools::subgraphFromNodes` is not used instead. On the other hand, first calculating the sets of nodes for each component seems wasteful if could also instead just iterate over all nodes `u` and erase `u` if `cc.componentOfNode(u) != largestCC->first`.